File: prototype/node/src/app/core/listings/image-format.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { sniffImageFormat } from './image-format.ts'

test('sniffs a PNG from its signature', () => {
  const bytes = Uint8Array.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0x00, 0x00])
  assert.equal(sniffImageFormat(bytes), 'png')
})

test('sniffs a JPEG from its signature', () => {
  const bytes = Uint8Array.from([0xff, 0xd8, 0xff, 0xe0, 0x00, 0x10])
  assert.equal(sniffImageFormat(bytes), 'jpeg')
})

test('sniffs a GIF87a from its signature', () => {
  const bytes = Uint8Array.from(Buffer.from('GIF87a rest of file', 'ascii'))
  assert.equal(sniffImageFormat(bytes), 'gif')
})

test('sniffs a GIF89a from its signature', () => {
  const bytes = Uint8Array.from(Buffer.from('GIF89a rest of file', 'ascii'))
  assert.equal(sniffImageFormat(bytes), 'gif')
})

test('sniffs a WebP from its RIFF/WEBP signature', () => {
  const bytes = Uint8Array.from(Buffer.from('RIFF____WEBPVP8 ', 'ascii'))
  assert.equal(sniffImageFormat(bytes), 'webp')
})

test('a RIFF file that is not WebP sniffs as nothing', () => {
  const bytes = Uint8Array.from(Buffer.from('RIFF____AVI rest', 'ascii'))
  assert.equal(sniffImageFormat(bytes), null)
})

test('SVG bytes sniff as nothing — text markup carries no magic bytes', () => {
  const bytes = Uint8Array.from(Buffer.from('<svg xmlns="http://www.w3.org/2000/svg"></svg>', 'utf8'))
  assert.equal(sniffImageFormat(bytes), null)
})

test('HTML bytes sniff as nothing', () => {
  const bytes = Uint8Array.from(Buffer.from('<html><body>evil</body></html>', 'utf8'))
  assert.equal(sniffImageFormat(bytes), null)
})

test('an empty buffer sniffs as nothing', () => {
  assert.equal(sniffImageFormat(Uint8Array.from([])), null)
})

test('a buffer shorter than any signature sniffs as nothing', () => {
  assert.equal(sniffImageFormat(Uint8Array.from([0x89, 0x50])), null)
})

File: prototype/rails/src/test/models/image_format_test.rb
require "test_helper"

class ImageFormatTest < ActiveSupport::TestCase
  test "sniffs a PNG from its signature" do
    assert_equal :png, ImageFormat.sniff("\x89PNG\r\n\x1a\n\x00\x00".b)
  end

  test "sniffs a JPEG from its signature" do
    assert_equal :jpeg, ImageFormat.sniff("\xff\xd8\xff\xe0\x00\x10".b)
  end

  test "sniffs a GIF87a from its signature" do
    assert_equal :gif, ImageFormat.sniff("GIF87a rest of file")
  end

  test "sniffs a GIF89a from its signature" do
    assert_equal :gif, ImageFormat.sniff("GIF89a rest of file")
  end

  test "sniffs a WebP from its RIFF/WEBP signature" do
    assert_equal :webp, ImageFormat.sniff("RIFF____WEBPVP8 ")
  end

  test "a RIFF file that is not WebP sniffs as nothing" do
    assert_nil ImageFormat.sniff("RIFF____AVI rest")
  end

  test "SVG bytes sniff as nothing — markup carries no magic bytes" do
    assert_nil ImageFormat.sniff('<svg xmlns="http://www.w3.org/2000/svg"></svg>')
  end

  test "HTML bytes sniff as nothing" do
    assert_nil ImageFormat.sniff("<html><body>evil</body></html>")
  end

  test "PDF bytes sniff as nothing" do
    assert_nil ImageFormat.sniff("%PDF-1.4\n")
  end

  test "a zip's local file header sniffs as nothing" do
    assert_nil ImageFormat.sniff("PK\x03\x04".b)
  end

  test "nil sniffs as nothing" do
    assert_nil ImageFormat.sniff(nil)
  end

  test "an empty string sniffs as nothing" do
    assert_nil ImageFormat.sniff("")
  end

  test "a buffer shorter than any signature sniffs as nothing" do
    assert_nil ImageFormat.sniff("\x89P".b)
  end
end
